web/packages/teleport/src/Discover/useDiscover.tsx

```diff
@@ -57,21 +65,38 @@ interface DiscoverContextState<T = any> {
   setResourceState: (value: T) => void;
   updateAgentMeta: (meta: AgentMeta) => void;
   views: View[];
+  emitErrorEvent(errorStr: string): void;
+  emitEvent(status: DiscoverEventStepStatus, eventName?: DiscoverEvent): void;
+  eventState: EventState;
 }
 
+type EventState = {
+  id: string;
+  currEventName: DiscoverEvent;
+  resource: DiscoverEventResource;
+};
+
 const discoverContext = React.createContext<DiscoverContextState>(null);
 
 export function DiscoverProvider<T = any>(
   props: React.PropsWithChildren<unknown>
 ) {
   const location = useLocation<{ entity: string }>();
+  const history = useHistory();
 
   const [currentStep, setCurrentStep] = useState(0);
   const [selectedResourceKind, setSelectedResourceKind] =
     useState<ResourceKind>(getKindFromString(location?.state?.entity));
   const [agentMeta, setAgentMeta] = useState<AgentMeta>();
   const [resourceState, setResourceState] = useState<T>();
 
+  const [eventState, setEventState] = useState<EventState>();
+  // Use the latest state for async use.
+  const ref = React.useRef<EventState>();
+  useEffect(() => {
+    ref.current = eventState;
+  }, [eventState]);
```

Comment:
Is this needed? 

Would having `eventState` as dependencies of the `useEffect`s and change the functions to `useCallback` with `eventState` as a dependency work instead?

Reply:
you're right, i learned that the cleanup func gets called on each call on the useeffect: https://github.com/gravitational/teleport/pull/20581/commits/55d631d8a8379eaea88996325c6cd1d87773f0b8